### PR TITLE
[Rust][Diagnostics] Add initial boilerplate for Rust diagnostic interface.

### PR DESCRIPTION
**Pull request by jroesch**
This PR exposes new diagnostics to Rust so we can reuse the better libraries for diagnostics rendering via Rust.

### CODE REVIEW

**Comment by jroesch** [comment]
cc @gussmith23 and @mwillsey

File: rust/compiler-ext/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "compiler-ext"
+version = "0.1.0"
+authors = ["Jared Roesch <jroesch@octoml.ai>"]
+edition = "2018"
+# TODO(@jroesch): would be cool to figure out how to statically link instead.
```
Comment (jroesch) [inline rust/compiler-ext/Cargo.toml:6]: I figured out how to do this.

File: rust/compiler-ext/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+fn test_fn2(message: tvm::runtime::string::String) -> Result<(), tvm::Error> {
+    println!("The message: {}", message);
+    Ok(())
+}
+
+tvm::export!(test_fn, test_fn2);
+
+#[no_mangle]
+fn compiler_ext_initialize() -> i32 {
```
Comment (tkonolige) [inline rust/compiler-ext/src/lib.rs:37]: This doesn't need an extern?

File: rust/tvm/src/ir/diagnostics/codespan.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
+use std::collections::HashMap;
```
Comment (tkonolige) [inline rust/tvm/src/ir/diagnostics/codespan.rs:22]: This file could use some comments

File: rust/compiler-ext/src/lib.rs (hunk 2 above)
Comment (jroesch) [reply rust/compiler-ext/src/lib.rs:37]: I think extern just sets symbol visibility, seems to have worked.

Comment (jroesch) [reply rust/compiler-ext/src/lib.rs:37]: I added it.

File: cmake/modules/RustExt.cmake [hunk 4]
```diff
@@ -0,0 +1,26 @@
+if(USE_RUST_EXT AND NOT USE_RUST_EXT EQUAL OFF)
```
Comment (rkimball) [inline cmake/modules/RustExt.cmake:1]: We really need to stop overloading cmake bool and string. cmake is a simple thing and is easily confused. In the meantime us this instead. OFF is only one of the many valid values for cmake bool false.
```suggestion
if(USE_RUST_EXT AND NOT ${USE_RUST_EXT} MATCHES ${IS_FALSE_PATTERN})
```

**Review by rkimball — CHANGES_REQUESTED** [review]

File: cmake/modules/RustExt.cmake (hunk 4 above)
Comment (jroesch) [reply cmake/modules/RustExt.cmake:1]: I am not a great CMake user so just let me know what the best way to do things is. I just have no clue what the "right way" in CMake is.

Comment (tkonolige) [reply cmake/modules/RustExt.cmake:1]: I think `if(USE_RUST_EXT)` should be fine.

Comment (tkonolige) [reply cmake/modules/RustExt.cmake:1]: _if(<constant>)
True if the constant is 1, ON, YES, TRUE, Y, or a non-zero number. False if the constant is 0, OFF, NO, FALSE, N, IGNORE, NOTFOUND, the empty string, or ends in the suffix -NOTFOUND. Named boolean constants are case-insensitive. If the argument is not one of these constants, it is treated as a variable._

File: rust/tvm/src/ir/diagnostics/mod.rs [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+//   /*! \brief Display diagnostics in a given display format.
+//    *
+//    * A diagnostic renderer is responsible for converting the
+//    * raw diagnostics into consumable output.
+//    *
+//    * For example the terminal renderer will render a sequence
+//    * of compiler diagnostics to std::out and std::err in
+//    * a human readable form.
+//    */
```
Comment (rkimball) [inline rust/tvm/src/ir/diagnostics/mod.rs:134]:
```suggestion
/// \brief Display diagnostics in a given display format.
///   
/// A diagnostic renderer is responsible for converting the
/// raw diagnostics into consumable output.
///  
/// For example the terminal renderer will render a sequence
/// of compiler diagnostics to std::out and std::err in
/// a human readable form.
///
```

File: rust/tvm/src/ir/span.rs [hunk 6]
```diff
@@ -1,22 +1,75 @@
...
+//    /*!
+//     * \brief Get an SourceName for a given operator name.
+//     *  Will raise an error if the source name has not been registered.
+//     * \param name Name of the operator.
+//     * \return SourceName valid throughout program lifetime.
+//     */
+//    TVM_DLL static SourceName Get(const String& name);
+
+//    TVM_DEFINE_OBJECT_REF_METHODS(SourceName, ObjectRef, SourceNameNode);
+//  };
```
Comment (rkimball) [inline rust/tvm/src/ir/span.rs:50]: Fix this double commented

File: cmake/modules/RustExt.cmake [hunk 7]
```diff
@@ -0,0 +1,26 @@
+if(USE_RUST_EXT)
+    set(RUST_SRC_DIR "${CMAKE_SOURCE_DIR}/rust")
+    set(CARGO_OUT_DIR "${CMAKE_SOURCE_DIR}/rust/target")
+
+    if(USE_RUST_EXT STREQUAL "STATIC")
+        set(COMPILER_EXT_PATH "${CARGO_OUT_DIR}/release/libcompiler_ext.a")
+    elseif(USE_RUST_EXT STREQUAL "DYNAMIC")
+        set(COMPILER_EXT_PATH "${CARGO_OUT_DIR}/release/libcompiler_ext.so")
+    else()
+        message(FATAL_ERROR "invalid setting for RUST_EXT")
```
Comment (rkimball) [inline cmake/modules/RustExt.cmake:10]:
```suggestion
        message(FATAL_ERROR "invalid setting for USE_RUST_EXT")
```

File: CMakeLists.txt [hunk 8]
```diff
@@ -79,6 +79,7 @@ tvm_option(USE_ARM_COMPUTE_LIB "Build with Arm Compute Library" OFF)
 tvm_option(USE_ARM_COMPUTE_LIB_GRAPH_RUNTIME "Build with Arm Compute Library graph runtime" OFF)
 tvm_option(USE_TENSORRT_CODEGEN "Build with TensorRT Codegen support" OFF)
 tvm_option(USE_TENSORRT_RUNTIME "Build with TensorRT runtime" OFF)
+tvm_option(USE_RUST_EXT "Build with Rust based compiler extensions" OFF)
```
Comment (rkimball) [inline CMakeLists.txt:82]: The help string should probably say something about STATIC and DYNAMIC

**Review by rkimball — CHANGES_REQUESTED** [review]

**Comment by jroesch** [comment]
I think this one is now ready to roll.

**Review by rkimball — APPROVED** [review]

File: cmake/modules/LLVM.cmake [hunk 9]
```diff
@@ -16,7 +16,12 @@
 # under the License.
 
 # LLVM rules
-add_definitions(-DDMLC_USE_FOPEN64=0)
+# Due to LLVM debug symbols you can sometimes face linking issues on
+# certain compiler, platform combinations if you don't set NDEBUG.
+#
+# See https://github.com/imageworks/OpenShadingLanguage/issues/1069
+# for more discussion.
+add_definitions(-DDMLC_USE_FOPEN64=0 -DNDEBUG=1)
```
Comment (tkonolige) [inline cmake/modules/LLVM.cmake:24]: It doesn't seem like a good idea to define `NDEBUG` globally.

File: cmake/modules/RustExt.cmake [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+if(USE_RUST_EXT)
+    set(RUST_SRC_DIR "${CMAKE_SOURCE_DIR}/rust")
+    set(CARGO_OUT_DIR "${CMAKE_SOURCE_DIR}/rust/target")
+
+    if(USE_RUST_EXT STREQUAL "STATIC")
+        set(COMPILER_EXT_PATH "${CARGO_OUT_DIR}/release/libcompiler_ext.a")
+    elseif(USE_RUST_EXT STREQUAL "DYNAMIC")
+        set(COMPILER_EXT_PATH "${CARGO_OUT_DIR}/release/libcompiler_ext.so")
+    else()
+        message(FATAL_ERROR "invalid setting for USE_RUST_EXT")
```
Comment (tkonolige) [inline cmake/modules/RustExt.cmake:27]: Could you print what the valid settings are here.

File: cmake/modules/LLVM.cmake (hunk 9 above)
Comment (jroesch) [reply cmake/modules/LLVM.cmake:24]: LLVM doesn't link correctly if you don't read the issues.

Comment (tkonolige) [reply cmake/modules/LLVM.cmake:24]:
```suggestion
add_definitions(-DDMLC_USE_FOPEN64=0)
target_add_definitions(tvm PRIVATE NDEBUG=1)
```

File: rust/tvm-rt/src/array.rs [hunk 11]
```diff
@@ -81,6 +82,40 @@ impl<T: IsObjectRef> Array<T> {
...
+    pos: isize,
+    size: isize,
+}
+
+impl<T: IsObjectRef> Iterator for IntoIter<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.pos < self.size {
+            let item = self.array.get(self.pos).expect("should not fail");
```
Comment (adelbertc) [inline rust/tvm-rt/src/array.rs:96]: Even though exceptional, better error message here than "should not fail"

File: rust/tvm-rt/src/function.rs [hunk 12]
```diff
@@ -120,21 +121,31 @@ impl Function {
...
+                num_args as c_int,
+                &mut ret_val as *mut _,
+                &mut ret_type_code as *mut _,
+            )
+        };
+
+        if ret_code != 0 {
+            let raw_error = crate::get_last_error();
+            let error = match Error::from_raw_tvm(raw_error) {
+                Error::Raw(string) => Error::CallFailed(string),
```
Comment (adelbertc) [inline rust/tvm-rt/src/function.rs:138]: This is kinda weird innit, transforming one variant to another, is the `from_raw_tvm` parser not able to handle that? Maybe with a flag or something?

File: rust/tvm-rt/src/function.rs [hunk 13]
```diff
@@ -120,21 +121,31 @@ impl Function {
...
+            return Err(error);
+        }
 
         let rv = RetValue::from_tvm_value(ret_val, ret_type_code as u32);
         match rv {
             RetValue::ObjectHandle(object) => {
-                let optr = crate::object::ObjectPtr::from_raw(object as _).unwrap();
-                // println!("after wrapped call: {}", optr.count());
-                crate::object::ObjectPtr::leak(optr);
+                let optr = ObjectPtr::from_raw(object as _).unwrap();
```
Comment (adelbertc) [inline rust/tvm-rt/src/function.rs:147]: `.except()` with a good error message?

File: rust/tvm/src/bin/tyck.rs [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+#[derive(Debug, StructOpt)]
+#[structopt(name = "tyck", about = "Parse and type check a Relay program.")]
+struct Opt {
+    /// Input file
+    #[structopt(parse(from_os_str))]
+    input: PathBuf,
+}
+
+fn main() -> Result<()> {
+    codespan::init().expect("Rust based diagnostics");
```
Comment (adelbertc) [inline rust/tvm/src/bin/tyck.rs:38]: Error message

File: rust/tvm/src/bin/tyck.rs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    let opt = Opt::from_args();
+    println!("{:?}", &opt);
```
Comment (adelbertc) [inline rust/tvm/src/bin/tyck.rs:40]: Intentional? Should this be a proper log line

File: rust/tvm/src/bin/tyck.rs [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    let _module = match IRModule::parse_file(opt.input) {
+        Err(ir::module::Error::TVM(Error::DiagnosticError(_))) => return Ok(()),
+        Err(e) => {
+            return Err(e.into());
+        }
```
Comment (adelbertc) [inline rust/tvm/src/bin/tyck.rs:45]: I'm guessing `rustfmt` would turn this into a single line, maybe something I can PR

File: rust/tvm/src/ir/diagnostics/codespan.rs [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+                    end_pos,
+                }
+            }
+            _ => panic!(),
+        }
+    }
+}
+
+/// A mapping for all files in a source map to byte ranges.
+struct SpanToByteRange {
```
Comment (adelbertc) [inline rust/tvm/src/ir/diagnostics/codespan.rs:98]: Is this a place where we could integrating like tracing-style spans? (looking for PR ideas)

File: rust/tvm/src/ir/module.rs [hunk 18]
```diff
@@ -25,16 +29,20 @@ use crate::runtime::{external, Object, ObjectRef};
...
-use tvm_macros::Object;
+use super::source_map::SourceMap;
 
 // TODO(@jroesch): define type
 type TypeData = ObjectRef;
 type GlobalTypeVar = ObjectRef;
 
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("{0}")]
```
Comment (adelbertc) [inline rust/tvm/src/ir/module.rs:40]: Do we also want to use `#[source]` for the errors?

File: rust/compiler-ext/Cargo.toml [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+edition = "2018"
+
+[lib]
+crate-type = ["staticlib", "cdylib"]
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+tvm = { path = "../tvm", default-features = false, features = ["static-linking"] }
+log = "*"
```
[the remark this answers is not in the document]
Comment (jroesch) [reply rust/compiler-ext/Cargo.toml:31]: Yeah tracing is better, just haven't done it yet.

File: cmake/modules/LLVM.cmake (hunk 9 above)
Comment (jroesch) [reply cmake/modules/LLVM.cmake]: On second pass this requires refactoring our CMake files, put a TODO in about.

File: rust/tvm-rt/src/function.rs (hunk 12 above)
Comment (jroesch) [reply rust/tvm-rt/src/function.rs:138]: We don't know what happened, RawError is just a wrapper around the stack trace that TVM creates. In this case we know what failed was a function call so I'm adding more structure to the error in this case so we can check if its random exception vs. issue with a function call.

File: rust/tvm/src/ir/module.rs (hunk 18 above)
Comment (jroesch) [reply rust/tvm/src/ir/module.rs:40]: This kind of stuff would be good follow up work imo.

File: rust/tvm/src/bin/tyck.rs (hunk 14 above)
Comment (jroesch) [reply rust/tvm/src/bin/tyck.rs:38]: This code isn't really done yet still need changes from other branch.

**Review by adelbertc — APPROVED** [review]

**Review by junrushao1994 — APPROVED** [review]

**Pull request merged**